src/Moq/Evaluator.cs

```diff
@@ -33,7 +33,7 @@ public static Expression PartialEval(Expression expression, Func<Expression, boo
 		/// <returns>A new tree with sub-trees evaluated and replaced.</returns>
 		public static Expression PartialEval(Expression expression)
 		{
-			return PartialEval(expression, e => e.NodeType != ExpressionType.Parameter);
+			return PartialEval(expression, e => e.NodeType != ExpressionType.Parameter && !(e is MatchExpression));
```

Comment:
Instead of this change, perhaps it would be better to add a check right inside `Evaluator` that recognizes irreducible `Extension` nodes.